java/ql/src/semmle/code/java/dataflow/Bound.qll

```diff
@@ -2,17 +2,12 @@
  * Provides classes for representing abstract bounds for use in, for example, range analysis.
  */
 
-import java
-private import SSA
-private import RangeUtils
+private import internal.rangeanalysis.BoundSpecific
 
 private newtype TBound =
   TBoundZero() or
   TBoundSsa(SsaVariable v) { v.getSourceVariable().getType() instanceof IntegralType } or
-  TBoundExpr(Expr e) {
-    e.(FieldRead).getField() instanceof ArrayLengthField and
-    not exists(SsaVariable v | e = v.getAUse())
-  }
+  TBoundExpr(Expr e) { nonSsaVariableBoundedExpr(e) }
```

Comment:
I think this should keep the negative conjunct in the shared part, as it is important in relation to `TBoundSsa` that the two don't overlap.
```suggestion
  TBoundExpr(Expr e) { interestingExprBound(e) and not exists(SsaVariable v | e = v.getAUse()) }
```